libbeat/cloudid/cloudid.go

```diff
@@ -33,6 +33,116 @@ import (
 
 const defaultCloudPort = "443"
 
+// CloudID encapsulates the encoded (i.e. raw) and decoded parts of Elastic Cloud ID.
+type CloudID struct {
+	id     string
+	esURL  string
+	kibURL string
+
+	auth     string
+	username string
+	password string
+}
+
+// NewCloudID constructs a new CloudID object by decoding the given cloud ID and cloud auth.
+func NewCloudID(cloudID string, cloudAuth string) (*CloudID, error) {
+	cid := CloudID{
+		id:   cloudID,
+		auth: cloudAuth,
+	}
+
+	if err := cid.decode(); err != nil {
+		return nil, err
```

Comment:
The error might report invalid base64 encodings and such. How about wrapping the error with an additional message saying: "invalid cloud id '%v'"?

Reply:
This might not be the best place to do the wrapping as `cid.decode()` is decoding the cloud id and cloud auth both and we probably don't want to expose the cloud auth value in logs. And if we just exposed the cloud id value in the error message, that might be misleading if the error was a result of decoding the cloud auth value.

But I get your point about wrapping the lower-level error (e.g. base64 encoding) with a higher-level message. So I've done that separately for cloud id and cloud auth inside the `decode()` method. See ca8241184.